### PR TITLE
[Vishruti Ranjan] iP

### PR DESCRIPTION
**Pull request by sevenseasofbri**

### CODE REVIEW

File: src/main/java/Duke.java [hunk 1]
```diff
@@ -1,10 +1,106 @@
...
+    }
+
+    public static void printList(){
+        System.out.println(YOUR_LIST);
+        for(int i=0; i<numItems; i++){
+            System.out.println("\t"+(i+1)+". "+ list[i]);
+        }
+        System.out.println(LINE_BREAK);
+    }
+    public static void addToList(String answer, TaskType taskType){
```
Comment (alstontham) [inline src/main/java/Duke.java]: Perhaps a more descriptive variable name for "answer" would be better here?

File: src/main/java/Duke.java [hunk 2]
```diff
@@ -1,10 +1,106 @@
...
+        while(true){
+            answer = in.nextLine();
+            if(answer.trim().equalsIgnoreCase("bye")){
+                break;
+            }
+            if(answer.trim().equalsIgnoreCase("list")){
+                printList();
+                continue;
+            }
+            if(answer.contains("done")){
```
Comment (alstontham) [inline src/main/java/Duke.java]: These nested if-statements were a bit hard to follow, but good job for attempting to handle possible errors!

**Review by alstontham — COMMENTED** [review]
Overall your code looks great and easy to follow! The only part I had some trouble following was the nested if-statements under the "answer.contains("done")" section. I don't think that sort of error handling is required for this Level but good job for incorporating it!

File: src/main/java/Duke.java [hunk 3]
```diff
@@ -1,10 +1,106 @@
...
+                answer = answer.trim();
+                String[] words = answer.split(" ");
+                if(words.length>1) {
```
Comment (judowha) [inline src/main/java/Duke.java]: for this part of condition statements, it has more than 3 levels of indentation. So may create some subfunctions to reduce the number of indentations.

File: src/main/java/Duke.java (hunk 1 above)
Comment (sevenseasofbri) [reply src/main/java/Duke.java]: I see, will incorporate the necessary changes! Thanks for your review :D

File: src/main/java/Duke.java (hunk 2 above)
Comment (sevenseasofbri) [reply src/main/java/Duke.java]: Yes, I will try and refactor the statements.

File: src/main/java/Duke.java (hunk 3 above)
Comment (sevenseasofbri) [reply src/main/java/Duke.java]: Got it! I will refactor to make it more readable. Thank you!

File: src/main/java/Duke.java [hunk 4]
```diff
@@ -1,10 +1,106 @@
+import java.util.Scanner;
 public class Duke {
+    public static Task[] list = new Task[100];
```
Comment (alstontham) [inline src/main/java/Duke.java]: For a collection of objects, you might want to consider using the plural form. So here instead of calling the array "list" you could call it "tasks" since it is a collection of tasks.

File: src/main/java/Duke.java [hunk 5]
```diff
@@ -1,10 +1,106 @@
...
+                    if(valueToMarkDone>0 && valueToMarkDone<=numItems) {
+                        if(!list[valueToMarkDone-1].isDone){
+                            totalTasksDone++;
+                        }
+                        list[valueToMarkDone - 1].markAsDone();
+                        System.out.println("\tAwesome! I've marked this task as done:");
+                        System.out.println("\t["+list[valueToMarkDone - 1]);
+                        System.out.println("\tOnly "+(numItems-totalTasksDone)+" to go! ;)");
+                    }
+                    else{
```
Comment (alstontham) [inline src/main/java/Duke.java]: I noticed that for a few cases of if-else statements, you formatted it like:
}
else{
I think it would be better to format it like this: "} else {" with the two ends of the curly braces on the same line as "else".

**Review by alstontham — COMMENTED** [review]
Just informing you of a few more areas you could touch up on coding standards!

File: src/main/java/duke/Duke.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            index++;
+        }
+        System.out.println(LINE_BREAK);
+    }
+
+    private static void printMessage(String message){
+        System.out.println(LINE_BREAK);
+        System.out.println(message);
+        System.out.println(LINE_BREAK);
+    }
```
Comment (nishanthelango) [inline src/main/java/duke/Duke.java]: Try to standardize your spaces (e.g line 57: index = 0, line 59: "\t" + (index+1) + ". " + task, line 65: (String message) {

File: src/main/java/duke/DukeOutOfBoundsException.java [hunk 7]
```diff
@@ -0,0 +1,4 @@
+package duke;
+
+public class DukeOutOfBoundsException extends Exception{
+}
```
Comment (nishanthelango) [inline src/main/java/duke/DukeOutOfBoundsException.java]: Consider adding constructors for your Exception classes.

File: src/main/java/duke/Save.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            switch (taskType){
+            case "T":
+                textIn = textIn+ (taskType+" | "+done+" | "+task.description+System.lineSeparator());
+                break;
+            case "D":
+                textIn = textIn + (taskType+" | "+done+" | "+task.description+" | "+((Deadline) task).getBy()+System.lineSeparator());
+                break;
+            case "E":
+                textIn = textIn + (taskType+" | "+done+" | "+task.description+" | "+((Event) task).getAt()+System.lineSeparator());
+            }
```
Comment (nishanthelango) [inline src/main/java/duke/Save.java]: Consider adding a default clause here to comply with coding standards.

File: src/main/java/duke/Save.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        file = new File(directory+"/"+fileName);
+        try{
+            if(!file.exists()){
+                if(!file.getParentFile().exists()){
+                    file.getParentFile().mkdirs();
+                }
+                file.createNewFile();
+            }
+        }catch(IOException e){
+            System.out.println("\tSomething went wrong: "+ e.getMessage());
```
Comment (nishanthelango) [inline src/main/java/duke/Save.java]: Try to avoid deep nesting (arrowhead code).

File: src/main/java/duke/Duke.java [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+            + "|  _ \\ _   _| | _____ \n"
+            + "| | | | | | | |/ / _ \\\n"
+            + "| |_| | |_| |   <  __/\n"
+            + "|____/ \\__,_|_|\\_\\___|\n";
+
+    public static final String LINE_BREAK = "\t____________________________________________________________";
+    public static final String YOUR_LIST = "\t_______________________~Your List~__________________________";
+    public enum TaskType{
+        TODO , DEADLINE, EVENT, INVALID
+    }
```
Comment (nishanthelango) [inline src/main/java/duke/Duke.java:22]: Good use of enum!

File: src/main/java/duke/Duke.java (hunk 6 above)
Comment (sevenseasofbri) [reply src/main/java/duke/Duke.java]: Noted!

File: src/main/java/duke/DukeOutOfBoundsException.java (hunk 7 above)
Comment (sevenseasofbri) [reply src/main/java/duke/DukeOutOfBoundsException.java]: Okay, will do.

File: src/main/java/duke/Duke.java (hunk 10 above)
Comment (sevenseasofbri) [reply src/main/java/duke/Duke.java:22]: Thanks for the feedback! I'll apply the necessary changes. :))
